File: static/modules/reports.js
window.module_reports = {
  render: async (container) => {
    const reports = [
      { id: 'inventory-valuation', name: 'Inventory Valuation', icon: '📦', desc: 'Qty × unit price from latest PO, grouped by category' },
      { id: 'open-ecos', name: 'Open ECOs by Priority', icon: '🔄', desc: 'Draft/review ECOs sorted by priority with age' },
      { id: 'wo-throughput', name: 'WO Throughput', icon: '⚙️', desc: 'Work orders completed with cycle time analysis' },
      { id: 'low-stock', name: 'Low Stock Report', icon: '⚠️', desc: 'Items below reorder point with suggested orders' },
      { id: 'ncr-summary', name: 'NCR Summary', icon: '📋', desc: 'Open NCRs by severity and defect type' }
    ];

    let currentReport = null;
    let woDays = 30;

    function renderMenu() {
      container.innerHTML = `<div class="card mb-4">
        <h2 class="text-lg font-semibold mb-4">Reports</h2>
        <div class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-4">
          ${reports.map(r => `<div class="border rounded-lg p-4 hover:bg-blue-50 dark:hover:bg-gray-700 cursor-pointer transition-colors" onclick="window._reportRun('${r.id}')">
            <div class="text-2xl mb-2">${r.icon}</div>
            <div class="font-semibold">${r.name}</div>
            <div class="text-sm text-gray-500">${r.desc}</div>
          </div>`).join('')}
        </div>
      </div>`;
    }

    async function runReport(id) {
      currentReport = id;
      const params = id === 'wo-throughput' ? `?days=${woDays}` : '';
      container.innerHTML = '<div class="card"><p class="text-gray-400">Loading report...</p></div>';
      try {
        const res = await api('GET', 'reports/' + id + params);
        const data = res.data;
        let html = `<div class="card">
          <div class="flex justify-between items-center mb-4">
            <div class="flex items-center gap-3">
              <button class="btn btn-secondary" onclick="window._reportBack()">← Back</button>
              <h2 class="text-lg font-semibold">${reports.find(r=>r.id===id)?.name || id}</h2>
            </div>
            <button class="btn btn-secondary" onclick="window._reportExport('${id}')">📥 Export CSV</button>
          </div>`;

        if (id === 'inventory-valuation') {
          html += renderInventoryValuation(data);
        } else if (id === 'open-ecos') {
          html += renderOpenECOs(data);
        } else if (id === 'wo-throughput') {
          html += renderWOThroughput(data);
        } else if (id === 'low-stock') {
          html += renderLowStock(data);
        } else if (id === 'ncr-summary') {
          html += renderNCRSummary(data);
        }
        html += '</div>';
        container.innerHTML = html;
      } catch(e) {
        container.innerHTML = `<div class="card"><p class="text-red-500">${e.message}</p><button class="btn btn-secondary mt-2" onclick="window._reportBack()">← Back</button></div>`;
      }
    }

    function renderInventoryValuation(data) {
      let html = '';
      for (const g of (data.groups || [])) {
        html += `<h3 class="font-semibold mt-4 mb-2">${g.category} <span class="text-sm text-gray-500">(Subtotal: $${g.subtotal.toFixed(2)})</span></h3>`;
        html += `<div class="overflow-x-auto"><table class="w-full text-sm mb-2"><thead><tr class="border-b text-left text-gray-500">
          <th class="pb-1">IPN</th><th class="pb-1">Description</th><th class="pb-1 text-right">Qty</th><th class="pb-1 text-right">Unit Price</th><th class="pb-1 text-right">Subtotal</th>
        </tr></thead><tbody>`;
        for (const it of g.items) {
          html += `<tr class="border-b border-gray-100"><td class="py-1 font-mono">${it.ipn}</td><td class="py-1">${it.description}</td>
            <td class="py-1 text-right">${it.qty_on_hand}</td><td class="py-1 text-right">$${it.unit_price.toFixed(4)}</td>
            <td class="py-1 text-right font-medium">$${it.subtotal.toFixed(2)}</td></tr>`;
        }
        html += '</tbody></table></div>';
      }
      html += `<div class="mt-4 pt-4 border-t text-lg font-bold text-right">Grand Total: $${(data.grand_total||0).toFixed(2)}</div>`;
      return html;
    }

    function renderOpenECOs(data) {
      let html = `<div class="overflow-x-auto"><table class="w-full text-sm"><thead><tr class="border-b text-left text-gray-500">
        <th class="pb-2">ID</th><th class="pb-2">Title</th><th class="pb-2">Status</th><th class="pb-2">Priority</th><th class="pb-2">Created By</th><th class="pb-2 text-right">Age (Days)</th>
      </tr></thead><tbody>`;
      for (const e of data) {
        html += `<tr class="border-b border-gray-100"><td class="py-2 font-mono text-blue-600">${e.id}</td><td class="py-2">${e.title}</td>
          <td class="py-2">${badge(e.status)}</td><td class="py-2">${badge(e.priority)}</td>
          <td class="py-2">${e.created_by}</td><td class="py-2 text-right">${e.age_days}</td></tr>`;
      }
      html += '</tbody></table></div>';
      if (data.length === 0) html = '<p class="text-gray-400 text-center py-4">No open ECOs</p>';
      return html;
    }

    function renderWOThroughput(data) {
      let html = `<div class="flex gap-2 mb-4">
        ${[30,60,90].map(d => `<button class="btn ${woDays===d?'btn-primary':'btn-secondary'}" onclick="window._reportWODays(${d})">${d} Days</button>`).join('')}
      </div>`;
      html += `<div class="grid grid-cols-2 md:grid-cols-3 gap-4 mb-4">
        <div class="bg-blue-50 dark:bg-gray-700 rounded-lg p-4"><div class="text-2xl font-bold">${data.total_completed}</div><div class="text-sm text-gray-500">Completed</div></div>
        <div class="bg-green-50 dark:bg-gray-700 rounded-lg p-4"><div class="text-2xl font-bold">${data.avg_cycle_time_days}</div><div class="text-sm text-gray-500">Avg Cycle (Days)</div></div>
      </div>`;
      if (Object.keys(data.count_by_status||{}).length > 0) {
        html += `<h3 class="font-semibold mb-2">By Status</h3><table class="w-full text-sm"><tbody>`;
        for (const [s,c] of Object.entries(data.count_by_status)) {
          html += `<tr class="border-b border-gray-100"><td class="py-1">${badge(s)}</td><td class="py-1 text-right font-medium">${c}</td></tr>`;
        }
        html += '</tbody></table>';
      }
      return html;
    }

    function renderLowStock(data) {
      let html = `<div class="overflow-x-auto"><table class="w-full text-sm"><thead><tr class="border-b text-left text-gray-500">
        <th class="pb-2">IPN</th><th class="pb-2">Description</th><th class="pb-2 text-right">On Hand</th><th class="pb-2 text-right">Reorder Point</th><th class="pb-2 text-right">Suggested Order</th><th class="pb-2"></th>
      </tr></thead><tbody>`;
      for (const it of data) {
        html += `<tr class="border-b border-gray-100"><td class="py-2 font-mono">${it.ipn}</td><td class="py-2">${it.description}</td>
          <td class="py-2 text-right text-red-600 font-medium">${it.qty_on_hand}</td><td class="py-2 text-right">${it.reorder_point}</td>
          <td class="py-2 text-right font-medium">${it.suggested_order}</td>
          <td class="py-2"><button class="text-blue-600 text-xs hover:underline" onclick="navigate('procurement')">Create PO →</button></td></tr>`;
      }
      html += '</tbody></table></div>';
      if (data.length === 0) html = '<p class="text-green-600 text-center py-4">✅ All items above reorder point</p>';
      return html;
    }

    function renderNCRSummary(data) {
      let html = `<div class="grid grid-cols-2 md:grid-cols-3 gap-4 mb-4">
        <div class="bg-red-50 dark:bg-gray-700 rounded-lg p-4"><div class="text-2xl font-bold">${data.total_open}</div><div class="text-sm text-gray-500">Open NCRs</div></div>
        <div class="bg-blue-50 dark:bg-gray-700 rounded-lg p-4"><div class="text-2xl font-bold">${data.avg_resolve_days}</div><div class="text-sm text-gray-500">Avg Resolve (Days)</div></div>
      </div>`;
      if (Object.keys(data.by_severity||{}).length > 0) {
        html += `<h3 class="font-semibold mb-2">By Severity</h3><table class="w-full text-sm mb-4"><tbody>`;
        for (const [s,c] of Object.entries(data.by_severity)) {
          html += `<tr class="border-b border-gray-100"><td class="py-1">${badge(s)}</td><td class="py-1 text-right font-medium">${c}</td></tr>`;
        }
        html += '</tbody></table>';
      }
      if (Object.keys(data.by_defect_type||{}).length > 0) {
        html += `<h3 class="font-semibold mb-2">By Defect Type</h3><table class="w-full text-sm"><tbody>`;
        for (const [d,c] of Object.entries(data.by_defect_type)) {
          html += `<tr class="border-b border-gray-100"><td class="py-1">${d}</td><td class="py-1 text-right font-medium">${c}</td></tr>`;
        }
        html += '</tbody></table>';
      }
      return html;
    }

    window._reportRun = (id) => runReport(id);
    window._reportBack = () => renderMenu();
    window._reportExport = (id) => {
      const params = id === 'wo-throughput' ? `&days=${woDays}` : '';
      window.open('/api/v1/reports/' + id + '?format=csv' + params, '_blank');
    };
    window._reportWODays = (d) => { woDays = d; runReport('wo-throughput'); };

    renderMenu();
  }
};

File: internal/handlers/common/reports.go
package common

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"math"
	"net/http"
	"strconv"
	"time"
)

// --- Inventory Valuation ---

// InvValuationItem represents a single inventory valuation line item.
type InvValuationItem struct {
	IPN       string  `json:"ipn"`
	Desc      string  `json:"description"`
	Category  string  `json:"category"`
	QtyOnHand float64 `json:"qty_on_hand"`
	UnitPrice float64 `json:"unit_price"`
	Subtotal  float64 `json:"subtotal"`
	PORef     string  `json:"po_ref"`
}

// InvValuationGroup groups valuation items by category.
type InvValuationGroup struct {
	Category string             `json:"category"`
	Items    []InvValuationItem `json:"items"`
	Subtotal float64            `json:"subtotal"`
}

// InvValuationReport is the full inventory valuation report.
type InvValuationReport struct {
	Groups     []InvValuationGroup `json:"groups"`
	GrandTotal float64             `json:"grand_total"`
}

// ReportInventoryValuation handles the inventory valuation report endpoint.
func (h *Handler) ReportInventoryValuation(w http.ResponseWriter, r *http.Request) {
	rows, err := h.DB.Query(`
		SELECT i.ipn, COALESCE(i.description,''), COALESCE(i.mpn,''), i.qty_on_hand,
			COALESCE((SELECT pl.unit_price FROM po_lines pl JOIN purchase_orders po ON pl.po_id=po.id
				WHERE pl.ipn=i.ipn ORDER BY po.created_at DESC LIMIT 1), 0) as unit_price,
			COALESCE((SELECT pl.po_id FROM po_lines pl JOIN purchase_orders po ON pl.po_id=po.id
				WHERE pl.ipn=i.ipn ORDER BY po.created_at DESC LIMIT 1), '') as po_ref
		FROM inventory i ORDER BY i.ipn`)
	if err != nil {
		http.Error(w, `{"error":"`+err.Error()+`"}`, 500)
		return
	}
	defer rows.Close()

	catMap := map[string][]InvValuationItem{}
	var catOrder []string
	for rows.Next() {
		var item InvValuationItem
		var mpn string
		rows.Scan(&item.IPN, &item.Desc, &mpn, &item.QtyOnHand, &item.UnitPrice, &item.PORef)
		item.Subtotal = item.QtyOnHand * item.UnitPrice
		// Derive category from IPN prefix
		item.Category = IPNCategory(item.IPN)
		if _, ok := catMap[item.Category]; !ok {
			catOrder = append(catOrder, item.Category)
		}
		catMap[item.Category] = append(catMap[item.Category], item)
	}

	report := InvValuationReport{}
	for _, cat := range catOrder {
		items := catMap[cat]
		grp := InvValuationGroup{Category: cat, Items: items}
		for _, it := range items {
			grp.Subtotal += it.Subtotal
		}
		report.GrandTotal += grp.Subtotal
		report.Groups = append(report.Groups, grp)
	}

	if r.URL.Query().Get("format") == "csv" {
		WriteCSV(w, "inventory-valuation", []string{"IPN", "Description", "Category", "Qty On Hand", "Unit Price", "Subtotal", "PO Ref"}, func(cw *csv.Writer) {
			for _, g := range report.Groups {
				for _, it := range g.Items {
					cw.Write([]string{it.IPN, it.Desc, it.Category, fmt.Sprintf("%.2f", it.QtyOnHand), fmt.Sprintf("%.4f", it.UnitPrice), fmt.Sprintf("%.2f", it.Subtotal), it.PORef})
				}
			}
		})
		return
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(report)
}

// --- Open ECOs by Priority ---

// OpenECOItem represents an open ECO in the report.
type OpenECOItem struct {
	ID        string `json:"id"`
	Title     string `json:"title"`
	Status    string `json:"status"`
	Priority  string `json:"priority"`
	CreatedBy string `json:"created_by"`
	CreatedAt string `json:"created_at"`
	AgeDays   int    `json:"age_days"`
}

// ReportOpenECOs handles the open ECOs report endpoint.
func (h *Handler) ReportOpenECOs(w http.ResponseWriter, r *http.Request) {
	rows, err := h.DB.Query(`SELECT id, title, status, priority, created_by, created_at FROM ecos WHERE status IN ('draft','review') ORDER BY
		CASE priority WHEN 'critical' THEN 0 WHEN 'high' THEN 1 WHEN 'normal' THEN 2 WHEN 'low' THEN 3 ELSE 4 END, created_at`)
	if err != nil {
		http.Error(w, `{"error":"`+err.Error()+`"}`, 500)
		return
	}
	defer rows.Close()
	var items []OpenECOItem
	now := time.Now()
	for rows.Next() {
		var e OpenECOItem
		rows.Scan(&e.ID, &e.Title, &e.Status, &e.Priority, &e.CreatedBy, &e.CreatedAt)
		if t, err := time.Parse("2006-01-02 15:04:05", e.CreatedAt); err == nil {
			e.AgeDays = int(now.Sub(t).Hours() / 24)
		}
		items = append(items, e)
	}
	if items == nil {
		items = []OpenECOItem{}
	}

	if r.URL.Query().Get("format") == "csv" {
		WriteCSV(w, "open-ecos", []string{"ID", "Title", "Status", "Priority", "Created By", "Created At", "Age (Days)"}, func(cw *csv.Writer) {
			for _, e := range items {
				cw.Write([]string{e.ID, e.Title, e.Status, e.Priority, e.CreatedBy, e.CreatedAt, strconv.Itoa(e.AgeDays)})
			}
		})
		return
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(items)
}

// --- WO Throughput ---

// WOThroughputReport is the work order throughput report.
type WOThroughputReport struct {
	Days             int            `json:"days"`
	CountByStatus    map[string]int `json:"count_by_status"`
	TotalCompleted   int            `json:"total_completed"`
	AvgCycleTimeDays float64        `json:"avg_cycle_time_days"`
}

// ReportWOThroughput handles the work order throughput report endpoint.
func (h *Handler) ReportWOThroughput(w http.ResponseWriter, r *http.Request) {
	days := 30
	if d := r.URL.Query().Get("days"); d != "" {
		if v, err := strconv.Atoi(d); err == nil && (v == 30 || v == 60 || v == 90) {
			days = v
		}
	}
	since := time.Now().AddDate(0, 0, -days).Format("2006-01-02 15:04:05")

	rows, err := h.DB.Query(`SELECT status, started_at, completed_at FROM work_orders WHERE completed_at IS NOT NULL AND completed_at >= ?`, since)
	if err != nil {
		http.Error(w, `{"error":"`+err.Error()+`"}`, 500)
		return
	}
	defer rows.Close()

	report := WOThroughputReport{Days: days, CountByStatus: map[string]int{}}
	var totalCycleHours float64
	cycleCount := 0
	for rows.Next() {
		var status string
		var startedAt, completedAt *string
		rows.Scan(&status, &startedAt, &completedAt)
		report.CountByStatus[status]++
		report.TotalCompleted++
		if startedAt != nil && completedAt != nil {
			if st, err1 := time.Parse("2006-01-02 15:04:05", *startedAt); err1 == nil {
				if ct, err2 := time.Parse("2006-01-02 15:04:05", *completedAt); err2 == nil {
					totalCycleHours += ct.Sub(st).Hours()
					cycleCount++
				}
			}
		}
	}
	if cycleCount > 0 {
		report.AvgCycleTimeDays = math.Round(totalCycleHours/float64(cycleCount)/24*100) / 100
	}

	if r.URL.Query().Get("format") == "csv" {
		WriteCSV(w, "wo-throughput", []string{"Days", "Total Completed", "Avg Cycle Time (Days)"}, func(cw *csv.Writer) {
			cw.Write([]string{strconv.Itoa(report.Days), strconv.Itoa(report.TotalCompleted), fmt.Sprintf("%.2f", report.AvgCycleTimeDays)})
			cw.Write([]string{})
			cw.Write([]string{"Status", "Count"})
			for s, c := range report.CountByStatus {
				cw.Write([]string{s, strconv.Itoa(c)})
			}
		})
		return
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(report)
}

// --- Low Stock ---

// LowStockItem represents a low-stock inventory item.
type LowStockItem struct {
	IPN            string  `json:"ipn"`
	Description    string  `json:"description"`
	QtyOnHand      float64 `json:"qty_on_hand"`
	ReorderPoint   float64 `json:"reorder_point"`
	ReorderQty     float64 `json:"reorder_qty"`
	SuggestedOrder float64 `json:"suggested_order"`
}

// ReportLowStock handles the low stock report endpoint.
func (h *Handler) ReportLowStock(w http.ResponseWriter, r *http.Request) {
	rows, err := h.DB.Query(`SELECT ipn, COALESCE(description,''), qty_on_hand, reorder_point, reorder_qty FROM inventory WHERE qty_on_hand < reorder_point AND reorder_point > 0 ORDER BY (reorder_point - qty_on_hand) DESC`)
	if err != nil {
		http.Error(w, `{"error":"`+err.Error()+`"}`, 500)
		return
	}
	defer rows.Close()
	var items []LowStockItem
	for rows.Next() {
		var it LowStockItem
		rows.Scan(&it.IPN, &it.Description, &it.QtyOnHand, &it.ReorderPoint, &it.ReorderQty)
		it.SuggestedOrder = it.ReorderQty
		if it.SuggestedOrder == 0 {
			it.SuggestedOrder = it.ReorderPoint - it.QtyOnHand
		}
		items = append(items, it)
	}
	if items == nil {
		items = []LowStockItem{}
	}

	if r.URL.Query().Get("format") == "csv" {
		WriteCSV(w, "low-stock", []string{"IPN", "Description", "Qty On Hand", "Reorder Point", "Reorder Qty", "Suggested Order"}, func(cw *csv.Writer) {
			for _, it := range items {
				cw.Write([]string{it.IPN, it.Description, fmt.Sprintf("%.2f", it.QtyOnHand), fmt.Sprintf("%.2f", it.ReorderPoint), fmt.Sprintf("%.2f", it.ReorderQty), fmt.Sprintf("%.2f", it.SuggestedOrder)})
			}
		})
		return
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(items)
}

// --- NCR Summary ---

// NCRSummaryReport is the NCR summary report.
type NCRSummaryReport struct {
	BySeverity     map[string]int `json:"by_severity"`
	ByDefectType   map[string]int `json:"by_defect_type"`
	TotalOpen      int            `json:"total_open"`
	AvgResolveDays float64        `json:"avg_resolve_days"`
}

// ReportNCRSummary handles the NCR summary report endpoint.
func (h *Handler) ReportNCRSummary(w http.ResponseWriter, r *http.Request) {
	report := NCRSummaryReport{BySeverity: map[string]int{}, ByDefectType: map[string]int{}}

	// Open NCRs
	rows, err := h.DB.Query(`SELECT COALESCE(severity,'unknown'), COALESCE(defect_type,'unknown') FROM ncrs WHERE status NOT IN ('closed','resolved')`)
	if err != nil {
		http.Error(w, `{"error":"`+err.Error()+`"}`, 500)
		return
	}
	defer rows.Close()
	for rows.Next() {
		var sev, dt string
		rows.Scan(&sev, &dt)
		report.BySeverity[sev]++
		report.ByDefectType[dt]++
		report.TotalOpen++
	}

	// Avg resolve time from resolved NCRs
	rrows, err := h.DB.Query(`SELECT created_at, resolved_at FROM ncrs WHERE resolved_at IS NOT NULL`)
	if err == nil {
		defer rrows.Close()
		var totalHours float64
		count := 0
		for rrows.Next() {
			var ca string
			var ra *string
			rrows.Scan(&ca, &ra)
			if ra != nil {
				if ct, e1 := time.Parse("2006-01-02 15:04:05", ca); e1 == nil {
					if rt, e2 := time.Parse("2006-01-02 15:04:05", *ra); e2 == nil {
						totalHours += rt.Sub(ct).Hours()
						count++
					}
				}
			}
		}
		if count > 0 {
			report.AvgResolveDays = math.Round(totalHours/float64(count)/24*100) / 100
		}
	}

	if r.URL.Query().Get("format") == "csv" {
		WriteCSV(w, "ncr-summary", []string{"Metric", "Value"}, func(cw *csv.Writer) {
			cw.Write([]string{"Total Open", strconv.Itoa(report.TotalOpen)})
			cw.Write([]string{"Avg Resolve Days", fmt.Sprintf("%.2f", report.AvgResolveDays)})
			cw.Write([]string{})
			cw.Write([]string{"Severity", "Count"})
			for k, v := range report.BySeverity {
				cw.Write([]string{k, strconv.Itoa(v)})
			}
			cw.Write([]string{})
			cw.Write([]string{"Defect Type", "Count"})
			for k, v := range report.ByDefectType {
				cw.Write([]string{k, strconv.Itoa(v)})
			}
		})
		return
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(report)
}

// --- Helpers ---

// IPNCategory extracts the category prefix from an IPN.
func IPNCategory(ipn string) string {
	parts := SplitIPN(ipn)
	if len(parts) > 0 {
		return parts[0]
	}
	return "Other"
}

// SplitIPN splits an IPN by hyphens.
func SplitIPN(ipn string) []string {
	var result []string
	current := ""
	for _, c := range ipn {
		if c == '-' {
			if current != "" {
				result = append(result, current)
			}
			current = ""
		} else {
			current += string(c)
		}
	}
	if current != "" {
		result = append(result, current)
	}
	return result
}

// WriteCSV writes CSV data with the given headers and content function.
func WriteCSV(w http.ResponseWriter, name string, headers []string, writeFn func(*csv.Writer)) {
	w.Header().Set("Content-Type", "text/csv")
	w.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=%s.csv", name))
	cw := csv.NewWriter(w)
	cw.Write(headers)
	writeFn(cw)
	cw.Flush()
}
